File: problems/0207-course-schedule/python/solution.py
def canFinish(numCourses, prerequisites):
    """Course Schedule — https://leetcode.com/problems/course-schedule/"""
    graph = [[] for _ in range(numCourses)]
    for course, pre in prerequisites:
        graph[course].append(pre)

    state = [0] * numCourses  # 0 = unvisited, 1 = visiting, 2 = done

    def dfs(c):
        if state[c] == 1:
            return False  # back edge -> cycle
        if state[c] == 2:
            return True
        state[c] = 1
        for pre in graph[c]:
            if not dfs(pre):
                return False
        state[c] = 2
        return True

    return all(dfs(c) for c in range(numCourses))

File: problems/0207-course-schedule/go/solution.go
package solution

// Course Schedule — https://leetcode.com/problems/course-schedule/
func canFinish(numCourses int, prerequisites [][]int) bool {
	graph := make([][]int, numCourses)
	for _, p := range prerequisites {
		graph[p[0]] = append(graph[p[0]], p[1])
	}
	state := make([]int, numCourses) // 0 unvisited, 1 visiting, 2 done
	var dfs func(c int) bool
	dfs = func(c int) bool {
		if state[c] == 1 {
			return false // back edge -> cycle
		}
		if state[c] == 2 {
			return true
		}
		state[c] = 1
		for _, pre := range graph[c] {
			if !dfs(pre) {
				return false
			}
		}
		state[c] = 2
		return true
	}
	for c := 0; c < numCourses; c++ {
		if !dfs(c) {
			return false
		}
	}
	return true
}
